Use Fastlane on CI to run the tests and build the app

mokagio opened the pull request:
Fixes #869.

Based on top of #874 because otherwise the App Store target won't build.

### Test
Nothing to test, if CI is green we're good.

### Review

Only one developer required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.

mokagio committed:
Extract some parameters in the Fastfile that we'll reuse soon

mokagio committed:
Add Fastlane lane for running the unit tests

mokagio committed:
Use Fastlane test lane in CI instead of xcodebuild in CI

mokagio committed:
Add Fastlane lane to build the App Store target

mokagio committed:
Build App Store target in CI using Fastlane

mokagio committed:
Add SearchQuery+Simplenote.swift to the App Store target

mokagio committed:
Use Xcode 12.4 on CircleCI

This is aimed at trying to fix the unit tests hanging after completing,
but it still doesn't hurt to run on the latest stable Xcode (usually).

mokagio committed:
Update Fastlane to latest (2.174.0) version

Maintenance update just because I noticed we were a few minor versions
behind.

mokagio commented on .circleci/config.yml:
```diff
@@ -28,47 +28,22 @@ jobs:
-                COMPILER_INDEX_STORE_ENABLE=NO \
-                CODE_SIGN_IDENTITY="" \
-                CODE_SIGNING_REQUIRED=NO
+          command: bundle exec fastlane test
```
Love how less code there's in the CircleCI layer. Would/Will make it simpler to migrate to a different provides.

AliSoftware replied:
<img alt="nod" src="http://static.skaip.org/img/emoticons/180x180/f6fcff/nod.gif" width="40" />

mokagio commented on .circleci/config.yml:
```diff
@@ -28,47 +28,22 @@ jobs:
     executor:
       name: ios/default
-      xcode-version: "12.0"
+      xcode-version: "12.4.0"
```
Did this just to get on a more up to date version.

mokagio commented on Gemfile.lock:
```diff
@@ -117,10 +118,11 @@ GEM
-    fastimage (2.2.1)
-    fastlane (2.171.0)
+    fastimage (2.2.2)
+    fastlane (2.174.0)
```
Before doing any work with Fastlane, better to get on the latest version 😄

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -13,6 +13,20 @@ ENV["DOWNLOAD_METADATA"]="./fastlane/download_metadata.swift"
 
 USER_ENV_FILE_PATH = File.join(Dir.home, '.simplenotemacos-env.default')
 
+workspace = 'Simplenote.xcworkspace'
+internal_scheme = 'Simplenote'
+app_store_scheme = 'Simplenote-AppStore'
+build_folder = 'build'
```
nit: since those are constants, you should scream them in all-uppercase. 👮‍♂️  #manual-rubocoping

```suggestion
WORKSPACE = 'Simplenote.xcworkspace'
INTERNAL_SCHEME = 'Simplenote'
APP_STORE_SCHEME = 'Simplenote-AppStore'
BUILD_FOLDER = 'build'
```

(+ then of course update the references in rest of code)

mokagio replied:
I DIDN'T KNOW YOU HAD TO SCREAM WHEN DEFINING CONSTANTS... Although, I saw it around enough times that I should have figured it out 😅 .

Would they remain uppercase if they became global values? That's a change I'm considering doing later while extracting some logic in a function.

```ruby
@WORKSPACE = ...

# or 

@workspace = ...
```

AliSoftware replied:
Global **variables** should not be all uppercase. The convention-over-configuration rule in ruby about this is that all-uppercase means the value should be considered **constant** — like if you `.freeze` it (in fact it's also good practice to `.freeze` constants, even though I admit I always forget).

The only reasonable reason to move those to be global **variables** instead is if you really need to make those values be **variable**, ie you plan to change their value during the script. And if you do, that means they won't be constant anymore so shouldn't be all-uppercase anymore.

That being said, global variables are frown upon in ruby, like they are in many other languages — even if not as bad as class variables (I even think that rubocop considers `@@classvar` usage to be a big warning if not an error), but still not ideal. And I can't really see a reason why you'd need those values in your script in particular (workspace name and scheme name etc) to be (global) **variables** rather than **constants** in that case anyway?

_(And also, `@` is for **instance** variables not global variables. I rarely use _global variables_ in ruby because it's generally bad practice anyway, but I _think_ it's supposed to use the `$` prefix? Would need to Google that to be sure though)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -13,6 +13,20 @@ ENV["DOWNLOAD_METADATA"]="./fastlane/download_metadata.swift"
+# On CI, it's useful to skip code signing unless necessary to avoid spending
+# time syncing the certificates and profiles with match. And while not
+# neccessary, it doesn't hurt on a dev machine either.
+skip_code_signing_xcargs = [
+  'COMPILER_INDEX_STORE_ENABLE=NO',
+  "CODE_SIGN_IDENTITY=''",
+  'CODE_SIGNING_REQUIRED=NO',
+]
```
I made it possible to pass a `Hash` to `xcargs` a while ago (https://github.com/fastlane/fastlane/pull/8051) so that it shell-escapes everything for you and makes it more readable; we should use this here 😇 

```suggestion
skip_code_signing_xcargs = {
  COMPILER_INDEX_STORE_ENABLE: 'NO',
  CODE_SIGN_IDENTITY: '',
  CODE_SIGNING_REQUIRED: 'NO',
}
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -29,6 +43,30 @@ before_all do
+    workspace: workspace,
+    scheme: internal_scheme,
+    clean: true,
+    xcargs: skip_code_signing_xcargs.join(' ')
```
Join is handled by https://github.com/fastlane/fastlane/pull/8051 already if you pass a `Hash` as per previous comment/suggestion.

```suggestion
    xcargs: skip_code_signing_xcargs
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -29,6 +43,30 @@ before_all do
+    output_directory: build_folder,
+    # Skip archive so we don't need to worry about code signing
+    skip_archive: true,
+    xcargs: skip_code_signing_xcargs.join(' ')
```
Ditto.
```suggestion
    xcargs: skip_code_signing_xcargs
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -176,8 +214,8 @@ end
+      workspace: workspace,
+      scheme: internal_scheme,
       clean: true,
       output_directory: "./build/")
```
Combine some [Han Solo](https://github.com/woocommerce/woocommerce-android/pull/3592#pullrequestreview-592833189) (reuse the constant) + me playing human rubocop (closing parentheses alignment), and you get:
```suggestion
      output_directory: BUILD_FOLDER
    )
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -29,6 +43,30 @@ before_all do
+lane :test do | options |
+  run_tests(
+    workspace: workspace,
+    scheme: internal_scheme,
```
Even if that's the default configuration set up on the project (is it?), I think it's better to be explicit here anyway:
```suggestion
    scheme: internal_scheme,
    configuration: 'Debug',
```

AliSoftware reviewed (commented):
Looking good in general, but left a couple of nits that would be nice to be addressed 🙂

AliSoftware commented:
Side note: is it intentional that this repo is not configured to "prevent PR merge until the PR gets at least one approval"?
(Is it because there are not enough people working on SN so it would slow down the process too much?)

mokagio commented:
@AliSoftware thank you for the review and sorry I have't been able to follow up on your great suggestions yet.

> Side note: is it intentional that this repo is not configured to "prevent PR merge until the PR gets at least one approval"?

🤔 maybe it was a glitch, or someone enabled it in the meantime? Anyways it's enabled now.

![image](https://user-images.githubusercontent.com/1218433/109129416-bc27d100-77a4-11eb-9eda-1f6c30c0ad12.png)

mokagio committed:
Use UPPERCASE for new constants in Fastfile

Thanks @AliSoftware

See
https://github.com/Automattic/simplenote-macos/pull/870#discussion_r578371300

mokagio committed:
Use hash instead of array for xcargs in Fastfile

Thanks @AliSoftware

See
https://github.com/Automattic/simplenote-macos/pull/870#discussion_r578376864

mokagio committed:
Replace a hardcoded build folder definition with BUILD_FOLDER

Thanks @AliSoftware.

See
https://github.com/Automattic/simplenote-macos/pull/870#discussion_r578378491

mokagio committed:
Explicitly set build configuration for test lane in Fastfile

Thanks @AliSoftware.

See
https://github.com/Automattic/simplenote-macos/pull/870#discussion_r578381470

mokagio commented:
Everything should be addressed now, thanks @AliSoftware.
